### PR TITLE
Port RecentListens to Typescript

### PR DESCRIPTION
**Pull request by ishaanshah**
# Description

We're moving over to Typescript for our frontend code. This pull request migrates the profile.jsx file to typescript.

We also do some linting fixes here.

# Problem
The LB frontend code is very hard to read or modify because it's complex and javascript provides no guarantee of what is available where etc.

# Solution
Use typescript instead, which helps us have types on react components and functions and catch errors at compile time.

The linter has also caught a few actual bugs too.

### CODE REVIEW

File: listenbrainz/webserver/static/js/src/types.d.ts [hunk 1]
```diff
@@ -31,6 +33,47 @@ declare interface ModalProps {
...
+  save_url?: string;
+  spotify: {
+    access_token?: string;
+    permission?: string;
+  };
+  user: User;
+  web_sockets_server_url: string;
+}
+
+declare interface RecentListensState {
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/types.d.ts:61]: I'd like to keep component specific props and states in the component file itself.

* We shouldn't need to import these types much anyways other than tests. 
* It is gonna be easier to read the props etc if they're in the file vs having to open a file which will potentially get huge if we put ALL types in it

File: listenbrainz/webserver/static/js/src/LastFMImporter.tsx [hunk 2]
```diff
@@ -90,23 +90,7 @@ document.addEventListener("DOMContentLoaded", () => {
...
-    <LastFmImporter
-      user={user}
-      profileUrl={profile_url}
-      apiUrl={api_url}
-      lastfmApiKey={lastfm_api_key}
-      lastfmApiUrl={lastfm_api_url}
-    />,
-    domContainer
-  );
+  ReactDOM.render(<LastFmImporter {...reactProps} />, domContainer);
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/LastFMImporter.tsx:95]: Let's keep the props camelCase here.

File: listenbrainz/webserver/static/js/src/RecentListens.tsx [hunk 3]
```diff
@@ -35,77 +45,57 @@ class RecentListens extends React.Component {
...
     }
-  }
+  };
 
-  connectWebsockets() {
-    this._socket = io.connect(this.props.web_sockets_server_url);
-    this._socket.on("connect", () => {
-      console.debug("Connected to websocket!");
-      switch (this.state.mode) {
+  connectWebsockets = () => {
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/RecentListens.tsx:68]: Let's add return types to these functions as well.

File: listenbrainz/webserver/static/js/src/types.d.ts [hunk 4]
```diff
@@ -31,6 +35,46 @@ declare interface ModalProps {
   onClose(event: React.MouseEvent<HTMLButtonElement>): void;
 }
 
+declare interface RecentListensProps {
+  api_url: string;
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/types.d.ts:39]: Let's keep the props camelCase.

File: listenbrainz/webserver/static/js/src/types.d.ts [hunk 5]
```diff
@@ -7,12 +11,16 @@ declare type SpotifyPlayDirection = "up" | "down" | "hidden";
...
 
+declare type AlertType = "danger" | "warning" | "success";
+
+declare type User = {
+  id?: string;
+  name: string;
+  auth_token?: string;
+};
+
 declare interface ImporterProps {
```
Comment (ishaanshah) [inline listenbrainz/webserver/static/js/src/types.d.ts:22]: Has been changed in #793

File: listenbrainz/webserver/static/js/src/SpotifyPlayer.tsx [hunk 6]
```diff
@@ -141,7 +141,6 @@ export default class SpotifyPlayer extends React.Component<
         }
         this.handleWarning("Could not find track on Spotify");
         this.playNextTrack();
-        return;
```
Comment (ishaanshah) [inline listenbrainz/webserver/static/js/src/SpotifyPlayer.tsx:144]: Extra return statements, removed by the linter. I checked if player was working as expected and the alert bug didn't appear. This should be safe to remove.

**Pull request merged**